File: benchmarks/java/BinaryTree.java
package benchmarks.java;


import benchmarkhandle.*;
import benchmarks.BaseBenchmark;


public class BinaryTree extends BaseBenchmark {

    private final static int minDepth = 8;


    public static void main(String[] args) {
        BinaryTree binaryTree = new BinaryTree();
        BenchmarkHandler benchmarkHandler = new BenchmarkHandler();
        BenchmarkMessage benchmarkMessage = new BenchmarkMessage(BenchmarkType.BINARYTREE, BenchmarkImplementation.JAVA, BenchmarkMetric.GARBAGE_COLLECTION);
        for (int i = 0; i < 10; i++) {
            benchmarkHandler.startMeasuringMetrics(benchmarkMessage, binaryTree, args);
        }
    }


    public void initAlgorithm(String[] args) {

        int n = 0;
        if (args.length > 0) n = Integer.parseInt(args[0]);

        // Comparison happen,bigger number will pick up
        int maxDepth = Math.max(minDepth + 2, n);
        //to find max depth +1
        int stretchDepth = maxDepth + 1;

        int check = (TreeNode.bottomUpTree(stretchDepth)).itemCheck();
        System.out.println("stretch tree of depth " + stretchDepth + "\t check: " + check);

        TreeNode longLivedTree = TreeNode.bottomUpTree(maxDepth);

        for (int depth = minDepth; depth <= maxDepth; depth += 2) {
            int iterations = 1 << ((maxDepth - depth) + minDepth);
            check = 0;

            for (int i = 1; i <= iterations; i++) {
                check += (TreeNode.bottomUpTree(depth)).itemCheck();
            }
            System.out.println(iterations + "\t trees of depth " + depth + "\t check: " + check);
        }
        System.out.println("long lived tree of depth " + maxDepth + "\t check: " + longLivedTree.itemCheck());
    }

    private static class TreeNode {
        private TreeNode left, right;

        TreeNode(TreeNode left, TreeNode right) {
            this.left = left;
            this.right = right;
        }

        //Here is the recursive loop
        private static TreeNode bottomUpTree(int depth) {
            if (depth > 0) {
                return new TreeNode(
                        bottomUpTree(depth - 1)
                        , bottomUpTree(depth - 1)
                );

            } else {
                return new TreeNode(null, null);
            }
        }

        private int itemCheck() {
            // if necessary deallocate here
            if (left == null) return 1;
            else return 1 + left.itemCheck() + right.itemCheck();
        }
    }
}

File: benchmarks/kotlinpure/BinaryTree.kt
package benchmarks.kotlinpure

import benchmarkhandle.*
import benchmarks.BaseBenchmark

object BinaryTree : BaseBenchmark() {

    private val minDepth = 8

    @JvmStatic
    fun main(args: Array<String>) {
        val binaryTree = BinaryTree
        val benchmarkHandler = BenchmarkHandler()
        val benchmarkMessage= BenchmarkMessage(BenchmarkType.BINARYTREE, BenchmarkImplementation.KOTLIN_PURE,BenchmarkMetric.MEMORY_CONSUMPTION)

        for (i in 0 until 10) {
            benchmarkHandler.startMeasuringMetrics(benchmarkMessage, binaryTree, args)
        }
    }

    override fun initAlgorithm(args: Array<String>) {
        var n = 0
        if (args.isNotEmpty()) n = Integer.parseInt(args[0])

        val maxDepth = if (n < minDepth + 2) minDepth + 2 else n
        val stretchDepth = maxDepth + 1


        var check = bottomUpTree(stretchDepth).itemCheck()
        println("stretch tree of depth $stretchDepth\t check: $check")

        val longLivedTree = bottomUpTree(maxDepth)

        var depth = minDepth
        while (depth <= maxDepth) {
            val iterations = 1 shl maxDepth - depth + minDepth
            check = 0

            for (i in 1 until iterations) {
                check += bottomUpTree(depth).itemCheck()
            }
            println("$iterations\t trees of depth $depth\t check: $check")
            depth += 2
        }
        println("long lived tree of depth " + maxDepth + "\t check: " + longLivedTree.itemCheck())
    }

    private fun bottomUpTree(depth: Int): TreeNode {
        return if (0 < depth) {
            TreeNode(bottomUpTree(depth - 1), bottomUpTree(depth - 1))
        } else TreeNode()
    }

    private data class TreeNode(
        val left: TreeNode? = null,
        val right: TreeNode? = null
    ) {
        fun itemCheck(): Int = if (null == left) 1 else 1 + left.itemCheck() + right!!.itemCheck()
    }
}